### DISCUSSION TITLE
Vscode: Extension Installation (Retains your extensions)

### DISCUSSION THREAD

**Discussion by kRHYME7** [prasanthrangan/hyprdots#526] [Show and tell]
Based on this #478 

This is an attempt to utilize *.vsix packages used by vscode. 
To avoid overwriting on themes  I suggest using a compressed version of the extensions WITH  Fav-Theme.vsix INCLUDED. 
(instructions at the bottom).
By this we can uncompress the Code_Extensions.tar.gz outside the ``` ~/.vscode/ ``` or as a temporary directory
then inside that folder, we can 
recursively search for *.vsix files and add it to ``` code --install-extension $Path-to-vsix --force. 
example: 
```
    
        find "${extension_dir}" -name "*.vsix" -print0 | while IFS= read -r -d '' vsix_file; do
            code --install-extension "$vsix_file"
        done
        rm -fr $extension_dir

```
By doing this we won't overwrite other extensions. 

Downsides:  Some re-compilations of the extensions to add the *.vsix files

Compiling  instructions: 
+ We need a package called  "vsce" to aquire the *.vsix of an extension, OR we can download the extension on web or on vscode app with *.vsix file included. 
you can get vsce in the AUR ``` paru -Sy vsce ```

+ Modifying the extensions package label into Fav-Theme.  ~~This is used by vscode to know what is the theme name, so you should edit the extension theme name to be the same to your $Fav_Theme 
Example: from "Hack The Box" to "Hackthebox"
On Json it should be at ``` contributes :  themes: label "Fav-Theme" ```
![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/9d63fb0d-00a3-4470-8c48-f01db638f0c3)
BUT This might not work so I don't know how ahha HELP me What Part should we edit to override it.~~ 
I don't know the right way haha

+ Generate *vsix :  cd to the extension directory with package.json located run ``` vsce package ``` 

+ Test: Test  if  the theme name of vscode changes after you  change your theme, you can check this inside ``` code ~/.config/Code/User/setting.json ``` If  your $Fav_Theme name is correctly displayed in "workbench.colorTheme" the theme of the vscode should take effect. 
![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/76d2adc0-153b-429f-86c2-485b129e9883)

+ Archiving: 
       Compress it to ``` Code_$Fav_Theme.tar.gz ```. The script also supports either "tar.gz" or "tar. xz ". 

I can start Writing this and share the progress to get some advice.
@prasanthrangan 
@T-Crypt 
@all

**Comment by kRHYME7** [comment]
My problem now is I don't know how to modify the extensions to set it to $Fav-Theme

**Reply by prasanthrangan** [reply]
Hey @kRHYME7 , i'll modify the control file to map vscode extension separately like this..

```
1|Catppuccin-Mocha|Catppuccin.catppuccin-vsc~Catppuccin Mocha|~/.config/swww/Catppuccin-Mocha/rain_world1.png
0|Catppuccin-Latte|Catppuccin.catppuccin-vsc~Catppuccin Latte|~/.config/swww/Catppuccin-Latte/reindeer_forest.jpg
0|Decay-Green|decaycs.decay~Cosmic Decay|~/.config/swww/Decay-Green/night_city.png
```

and update the restore script to install these 
```
code --install-extension decaycs.decay
```

so that we don't have to maintain it in this repo!

**Reply by kRHYME7** [reply]
This sounds on point, I needed to adjust the Theme Patch right? With this route, I should consider making the control file a dependency for checking the validity of the theme Repo. Did I get it? 

The issue we may face is the inconsistent extension IDs for the themes, so my guess is we really needed the control file for validation.

**Comment by prasanthrangan** [comment]
> This sounds on point, I needed to adjust the Theme Patch right? With this route, I should consider making the control file a dependency for checking the validity of the theme Repo. Did I get it?

yup, the themepatch should be updated and ignore the vscode extension part ... just get it as an optional input parameter...
I've updated the control structure, and its in themes_v2 branch as below -

in `/.config/hypr/theme.ctl`
```
1|Catppuccin-Mocha|Catppuccin.catppuccin-vsc~Catppuccin Mocha|~/.config/swww/Catppuccin-Mocha/rain_world1.png
0|Catppuccin-Latte|Catppuccin.catppuccin-vsc~Catppuccin Latte|~/.config/swww/Catppuccin-Latte/reindeer_forest.jpg
0|Decay-Green|decaycs.decay~Decayce|~/.config/swww/Decay-Green/night_city.png
0|Rose-Pine|mvllow.rose-pine~Rosé Pine|~/.config/swww/Rose-Pine/mountain_sunset.jpg
0|Tokyo-Night|enkia.tokyo-night~Tokyo Night Storm|~/.config/swww/Tokyo-Night/tokyo_pink.png
0|Material-Sakura|mvllow.rose-pine~Rosé Pine Dawn|~/.config/swww/Material-Sakura/material_pastel_5.jpg
0|Graphite-Mono|StepanVanzuriak.mono~mono dark|~/.config/swww/Graphite-Mono/wall_paint.jpg
0|Cyberpunk-Edge|JWSandeman.cyberpunk2077-theme~cyberpunk2077|~/.config/swww/Cyberpunk-Edge/lucy_1.jpg
0|Frosted-Glass|msnilshartmann.blue-light~Blue Light Theme|~/.config/swww/Frosted-Glass/blue_desert.png
0|Gruvbox-Retro|jdinhlife.gruvbox~Gruvbox Dark Medium|~/.config/swww/Gruvbox-Retro/wall.jpg
```

so here in the 3rd column, its the extension id (is always unique) and the theme name (as an extension can have multiple themes in it)
Catppuccin.catppuccin-vsc~Catppuccin Mocha

and the script will install the `Catppuccin.catppuccin-vsc` extension and apply the `Catppuccin Mocha` vscode theme...

**Reply by kRHYME7** [reply]
I will try to modify it tomorrow. BTW  is the themev2 branch just a test branch right? I will merge the Theme Patch with this branch first.

**Reply by prasanthrangan** [reply]
> I will try to modify it tomorrow. BTW is the themev2 branch just a test branch right? I will merge the Theme Patch with this branch first.

Hey no worries, I've updated testpatcher in `themev2`... please feel free to test it in the branch...
once everything looks good, I'll merge the branch into main!

**Comment by kRHYME7** [comment]
Will  use the script on themev2 branch and will adapt the syntax

Concerns:
+ What should be the naming schemes for the packaged files? Is it mandatory to follow $Fav-Theme? (in case dashes exist )
+ Does my system have issues because I got operation not permitted when I run tar without sudo root privileges.
+ Are the warning messages also exibits on your system while using code --install-extension? 
```
 (node:61586) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
(Use `code --trace-deprecation ...` to show where the warning was created) 
```

I will merge themev2 to the theme patcher PR to avoid conflicts.

**Reply by prasanthrangan** [reply]
> * What should be the naming schemes for the packaged files? Is it mandatory to follow $Fav-Theme? (in case dashes exist )

Yup, using `$Fav-Theme` is convenient when renaming all config files with the same theme name -

![image](https://github.com/prasanthrangan/hyprdots/assets/106020512/49988b33-a706-4f51-8fb8-40aed62e87ab)

> * Does my system have issues because I got operation not permitted when I run tar without sudo root privileges.

you don't need sudo permission as you are just using your home directory...

> * Are the warning messages also exibits on your system while using code --install-extension?

ya, they are just warnings, the extensions are getting installed without issues.

> I will merge themev2 to the theme patcher PR to avoid conflicts.

I've already merged themepatcher in themev2 branch... Its working fine as expected... please give it a try and let me know if it works!

I did create a theme for it as well - https://github.com/prasanthrangan/hyprdots-mod
you can pull the themev2 branch and use the themepatcher as below -
```
./themepatcher.sh "Synth-Wave" "https://github.com/prasanthrangan/hyprdots-mod" "robbowen.synthwave-vscode~SynthWave '84"
```

https://github.com/prasanthrangan/hyprdots-mod/assets/106020512/0c7f12a8-11f2-4a16-890e-44f07a860636

**Reply by kRHYME7** [reply]
> you don't need sudo permission as you are just using your home directory...

I guess something is wrong on my end, because I do get some permission ~~issues~~ Warnings . Thanks

~~Update:~~

~~It works without sudo on my other laptop So I guess this was an issue per system.~~


Conclusion: 
 
I tested your "Synth-Wave" And everything works, I guess this was something to do with files from the ones I've tested. For context, I'm testing using this script to @T-Crypt 's  themes, ( Looks very very awesome and clean too).


For @T-Crypt 

``` 
tar: MacOS/cinnamon/assets: Cannot utime: Operation not permitted
tar: MacOS/cinnamon/assets: Cannot change mode to rwxr-xr-x: Operation not permitted
tar: MacOS/cinnamon: Cannot utime: Operation not permitted
tar: MacOS/cinnamon: Cannot change mode to rwxr-xr-x: Operation not permitted
tar: MacOS/gnome-shell/assets: Cannot utime: Operation not permitted
tar: MacOS/gnome-shell/assets: Cannot change mode to rwxr-xr-x: Operation not permitted
tar: MacOS/gnome-shell/icons/scalable/actions: Cannot utime: Operation not permitted

....
...
Also on the other themes.
```
